### PR TITLE
Nfdi

### PR DESCRIPTION
**Pull request by koerberm**
A first implementation of a connection to the NFDI core storage; It currently supports single-file vector and raster data;

### CODE REVIEW

**Comment by koerberm** [comment]
[NFDI-Core-Storage-Data-Provider](https://trello.com/c/c3GGDaIo/446-nfdi-core-storage-data-provider)

File: Settings-default.toml [hunk 1]
```diff
@@ -33,7 +33,7 @@ path = "upload"
 [logging]
 # Minimum log level. Can be one of error, warn, info, debug, trace
 # or a more detailed spec. See https://docs.rs/flexi_logger/0.17.1/flexi_logger/struct.LogSpecification.html.
-log_spec = "info"
+log_spec = "debug"
```
Comment (ChristianBeilschmidt) [inline Settings-default.toml:36]: We should not change the default here.

File: Settings-test.toml [hunk 2]
```diff
@@ -9,6 +9,11 @@ password = "geoengine"
 [operators.gdal_source]
 raster_data_root_path = "../test_data/raster" # relative to sub crate directory for tests
 
+[logging]
+# Minimum log level. Can be one of error, warn, info, debug, trace
+# or a more detailed spec. See https://docs.rs/flexi_logger/0.17.1/flexi_logger/struct.LogSpecification.html.
+log_spec = "debug"
```
Comment (ChristianBeilschmidt) [inline Settings-test.toml:15]: Do we need `debug` logging here or is this just for own debug purposes?

File: services/Cargo.toml [hunk 3]
```diff
@@ -15,6 +15,7 @@ xml = ["quick-xml"]
 postgres = ["postgres-types", "bb8-postgres"]
 odm = ["pro"]
 gfbio = ["postgres", "geoengine-datatypes/postgres"]
+nfdi = []
```
Comment (ChristianBeilschmidt) [inline services/Cargo.toml:18]: I guess we should merge gfbio and nfdi as a flag. I suggest both is nfdi now.

File: services/Cargo.toml [hunk 4]
```diff
@@ -53,6 +54,7 @@ pwhash = "1.0"
 quick-xml = { version = "0.22", optional = true }
 rayon = "1.5"
 reqwest = { version = "0.11", features = ["json", "multipart", "stream"] }
+scienceobjectsdb_rust_api = { git = "https://github.com/ScienceObjectsDB/rust-api", branch="main" }
```
Comment (ChristianBeilschmidt) [inline services/Cargo.toml:57]: Can't we move this dependency to only be used if nfdi is enabled?
It seems to be an unstable dependency by now.

File: services/src/datasets/external/nfdi/mod.rs [hunk 5]
```diff
@@ -0,0 +1,1031 @@
...
+
+        OgrSourceDataset {
+            file_name: PathBuf::from(link),
+            layer_name,
+            data_type,
+            time,
+            default_geometry: None,
+            columns: Some(column_spec),
+            force_ogr_time_filter: false,
+            force_ogr_spatial_filter: false,
```
Comment (ChristianBeilschmidt) [inline services/src/datasets/external/nfdi/mod.rs:327]: Do filters work or do we have to force this?

File: services/src/datasets/external/nfdi/mod.rs [hunk 6]
```diff
@@ -0,0 +1,1031 @@
...
+                    _phantom: Default::default(),
+                    object_stub: self.object_stub.clone(),
+                };
+                Ok(Box::new(res))
+            }
+            DataType::SingleVectorFile(_) => Err(geoengine_operators::error::Error::InvalidType {
+                found: md.data_type.to_string(),
+                expected: "SingleRasterFile".to_string(),
+            }),
+        }
```
Comment (ChristianBeilschmidt) [inline services/src/datasets/external/nfdi/mod.rs:469]: For single files, the 15-minute-key should be fine. We have to find another approach for multiple files and large (chunked) ogr sources.

File: services/src/datasets/external/nfdi/mod.rs [hunk 7]
```diff
@@ -0,0 +1,1031 @@
...
+    //     let result: Vec<MultiPointCollection> = proc
+    //         .query(qr, &ctx)
+    //         .await
+    //         .unwrap()
+    //         .map(Result::unwrap)
+    //         .collect()
+    //         .await;
+    //     let sum: usize = result.iter().map(MultiPointCollection::len).sum();
+    //     println!("{}", sum);
+    // }
```
Comment (ChristianBeilschmidt) [inline services/src/datasets/external/nfdi/mod.rs:1030]: I guess these are up to implement

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: Settings-default.toml (hunk 1 above)
Comment (koerberm) [reply Settings-default.toml]: 8fb81dbed4ec76c5ed14d5584083f7938818d7fe

File: Settings-test.toml (hunk 2 above)
Comment (koerberm) [reply Settings-test.toml]: Was for local testing. Removed it. 8fb81dbed4ec76c5ed14d5584083f7938818d7fe

File: services/Cargo.toml (hunk 3 above)
Comment (koerberm) [reply services/Cargo.toml]: c5fa0fb1b47bcbd79d7efced816f5f360cbc0d39

File: services/Cargo.toml (hunk 4 above)
Comment (koerberm) [reply services/Cargo.toml]: d760b3962b50d6acba8502aa4527c7b00ba83b5d

File: services/src/datasets/external/nfdi/mod.rs (hunk 5 above)
Comment (koerberm) [reply services/src/datasets/external/nfdi/mod.rs:352]: Yes, I just tested it.

File: services/src/datasets/external/nfdi/mod.rs (hunk 6 above)
Comment (koerberm) [reply services/src/datasets/external/nfdi/mod.rs:494]: I think we need to discuss this with Marius

File: services/src/datasets/external/nfdi/mod.rs (hunk 7 above)
Comment (koerberm) [reply services/src/datasets/external/nfdi/mod.rs:1062]: Should I remove them? We could basically reuse them after we mocked the core-storage API

**Review by ChristianBeilschmidt — DISMISSED** [review]

**Review by ChristianBeilschmidt — APPROVED** [review]

**Pull request merged**